### PR TITLE
introduce dump-signedexchange tool

### PR DESCRIPTION
**Pull request by nyaxt**
PTAL

### CODE REVIEW

**Review by hajimehoshi — CHANGES_REQUESTED** [review]

File: go/signedexchange/cbor/decoder.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+func (d *Decoder) decodeTypedUInt() (Type, uint64, error) {
+	b, err := d.ReadByte()
+	if err != nil {
+		return TypeOther, 0, err
+	}
+
+	t := Type(b & MaskType)
+	ai := b & MaskAdditionalInformation
+	nfollow := 0
+	switch ai {
```
[the remark this answers is not in the document]
Comment (nyaxt) [reply go/signedexchange/cbor/decoder.go]: yes `ai` also used below.

File: go/signedexchange/signedexchange.go [hunk 2]
```diff
@@ -187,3 +261,65 @@ func WriteExchangeFile(w io.Writer, e *Exchange) error {
...
+		fmt.Fprintf(w, "    %s: %s\n", k, e.responseHeaders.Get(k))
+	}
+	fmt.Fprintf(w, "response:\n")
+	fmt.Fprintf(w, "  status: %d\n", e.responseStatus)
+	fmt.Fprintf(w, "  headers:\n")
+	for k, _ := range e.responseHeaders {
+		fmt.Fprintf(w, "    %s: %s\n", k, e.responseHeaders.Get(k))
+	}
+	fmt.Fprintf(w, "payload [%d bytes]:\n", len(e.payload))
+	_, _ = w.Write(e.payload)
```
[the remark this answers is not in the document]
Comment (nyaxt) [reply go/signedexchange/signedexchange.go]: Done.

File: go/signedexchange/signedexchange.go [hunk 3]
```diff
@@ -187,3 +261,65 @@ func WriteExchangeFile(w io.Writer, e *Exchange) error {
...
+	if err := mice.Decode(&payloadBuf, r, miHeaderValue); err != nil {
+		return nil, fmt.Errorf("Failed to mice decode payload: %v", err)
+	}
+	e.payload = payloadBuf.Bytes()
+
+	return e, nil
+}
+
+func (e *Exchange) PrettyPrint(w io.Writer) {
+	fmt.Fprintf(w, "request:\n")
```
[the remark this answers is not in the document]
Comment (nyaxt) [reply go/signedexchange/signedexchange.go]: Done.

File: go/signedexchange/cbor/decoder.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	return t, n, nil
+}
+
+func (d *Decoder) decodeUintOfType(expected Type) (uint64, error) {
+	t, n, err := d.decodeTypedUInt()
+	if err != nil {
+		return 0, err
+	}
+	if t != expected {
+		return 0, fmt.Errorf("Expected type %v, got type %v", expected, t)
```
[the remark this answers is not in the document]
Comment (nyaxt) [reply go/signedexchange/cbor/decoder.go]: Done.

File: go/signedexchange/cbor/decoder.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	return d.decodeUintOfType(TypeMap)
+}
+
+func (d *Decoder) decodeBytesOfType(expected Type) ([]byte, error) {
+	n, err := d.decodeUintOfType(expected)
+	if err != nil {
+		return nil, err
+	}
+	bs := make([]byte, n)
+	if _, err := d.r.Read(bs); err != nil {
```
[the remark this answers is not in the document]
Comment (nyaxt) [reply go/signedexchange/cbor/decoder.go]: Good catch!

File: go/signedexchange/signedexchange.go [hunk 6]
```diff
@@ -26,6 +27,14 @@ type Exchange struct {
 	payload []byte
 }
 
+var (
+	KeyMethod = []byte(":method")
+	KeyURL    = []byte(":url")
+	KeyStatus = []byte(":status")
+
+	ValueGet = []byte("GET")
```
[the remark this answers is not in the document]
Comment (nyaxt) [reply go/signedexchange/signedexchange.go]: Done.

File: go/signedexchange/signedexchange.go [hunk 7]
```diff
@@ -125,6 +168,37 @@ func (e *Exchange) encodeResponseHeaders(enc *cbor.Encoder) error {
...
+		if err != nil {
+			return fmt.Errorf("Failed to decode value bytestring: %s", err)
+		}
+		// TODO: add key/value str validation?
+
+		if bytes.Equal(key, KeyStatus) {
+			// TODO: add value str validation that it only contains [0-9]
+			e.responseStatus, err = strconv.Atoi(string(value))
+			if err != nil {
+				log.Printf("Failed to parse responseStatus: %q", value)
```
[the remark this answers is not in the document]
Comment (nyaxt) [reply go/signedexchange/signedexchange.go]: I'll leave this w/ FIXME comment for now. I think we want some channel to communicate `lint`-like msg that the signed-exchange is ill-formed.

**Comment by nyaxt** [comment]
OK. PTAL

File: go/signedexchange/mice/mice.go [hunk 8]
```diff
@@ -57,3 +58,33 @@ func Encode(w io.Writer, buf []byte, recordSize int) (string, error) {
...
+	if err := binary.Read(r, binary.BigEndian, &recordSize); err != nil {
+		return fmt.Errorf("Failed to read recordSize: %v", err)
+	}
+
+	proof := make([]byte, sha256.Size)
+	record := make([]byte, recordSize)
+	readFirstRecord := false
+	for {
+		if readFirstRecord {
+			if _, err := r.Read(proof); err != nil {
```
Comment (hajimehoshi) [inline go/signedexchange/mice/mice.go]: `io.ReadFull` ?

File: go/signedexchange/mice/mice.go [hunk 9]
```diff
@@ -57,3 +58,33 @@ func Encode(w io.Writer, buf []byte, recordSize int) (string, error) {
...
+				if err == io.EOF {
+					return nil
+				}
+				return fmt.Errorf("Failed to read proof: %v", err)
+			}
+		}
+		readFirstRecord = true
+		n, err := r.Read(record)
```
Comment (hajimehoshi) [inline go/signedexchange/mice/mice.go]: ditto

File: go/signedexchange/mice/mice.go [hunk 10]
```diff
@@ -57,3 +58,33 @@ func Encode(w io.Writer, buf []byte, recordSize int) (string, error) {
...
+		if err != nil {
+			return fmt.Errorf("Failed to read record: %v", err)
```
Comment (hajimehoshi) [inline go/signedexchange/mice/mice.go]: `mice: failed to read ...`

File: go/signedexchange/mice/mice.go [hunk 11]
```diff
@@ -57,3 +58,33 @@ func Encode(w io.Writer, buf []byte, recordSize int) (string, error) {
...
+		}
+		// TODO: verify integrity
+		if _, err = w.Write(record[:n]); err != nil {
+			return fmt.Errorf("Failed to write record: %v", err)
```
Comment (hajimehoshi) [inline go/signedexchange/mice/mice.go]: ditto

File: go/signedexchange/signedexchange.go [hunk 12]
```diff
@@ -187,3 +263,66 @@ func WriteExchangeFile(w io.Writer, e *Exchange) error {
...
+	var encodedCborLength [3]byte
+	if _, err := r.Read(encodedCborLength[:]); err != nil {
+		return nil, fmt.Errorf("signedexchange: Failed to read length header")
+	}
+	cborLength := int(encodedCborLength[0])<<16 |
+		int(encodedCborLength[1])<<8 |
+		int(encodedCborLength[2])
+
+	cborBytes := make([]byte, cborLength)
+	if _, err := r.Read(cborBytes); err != nil {
```
Comment (hajimehoshi) [inline go/signedexchange/signedexchange.go]: `io.ReadFull`

File: go/signedexchange/signedexchange.go [hunk 13]
```diff
@@ -187,3 +263,66 @@ func WriteExchangeFile(w io.Writer, e *Exchange) error {
 
 	return nil
 }
+
+func ReadExchangeFile(r io.Reader) (*Exchange, error) {
+	var encodedCborLength [3]byte
+	if _, err := r.Read(encodedCborLength[:]); err != nil {
```
Comment (hajimehoshi) [inline go/signedexchange/signedexchange.go]: `io.ReadFull`

**Review by hajimehoshi — CHANGES_REQUESTED** [review]
almost lgtm

**Comment by nyaxt** [comment]
PTAL

**Review by hajimehoshi — APPROVED** [review]
lgtm

**Comment by twifkak** [comment]
@nyaxt Still planning to merge this? I'd like to use it for tests in ampproject/amppackager.

File: go/signedexchange/signedexchange.go [hunk 14]
```diff
@@ -26,6 +27,14 @@ type Exchange struct {
 	payload []byte
```
Comment (twifkak) [inline go/signedexchange/signedexchange.go]: Would you mind exporting these fields as part of this change? Or I can open a PR if you prefer.

**Comment by nyaxt** [comment]
Sorry, I forgot to push the merge btn.

**Pull request merged**